### PR TITLE
Rust support

### PR DESCRIPTION
**Pull request by Agile86**
Hello everybody,

my name is [Oleh Dolhov](https://github.com/Agile86), my collegue name is [Vitaly Reshetyuk](https://github.com/revitalyr),

we are from Stroz Friedberg (AON) providing this PR to bring Rust-language support to Kaitai.

Right now we can pass almost all tests, except (4) tests, that's using features like:

1. _parent rewrite_

Rust doesn't support inheritance, so there is no common type.

2. _KStructUnit_, tests like `nav_parent_switch_cast.ksy`, `params_pass_array_struct.ksy`, `params_pass_struct.ksy`.

3. _AnyType_, test like `process_coerce_switch.ksy`.

Please, note, you need to clone [tests](https://github.com/Agile86/kaitai_struct_tests/commits/rust_basic_support_v2)  and [runtime/rust](https://github.com/Agile86/kaitai_rust/tree/rust_basic_support_v2) exactly `rust_basic_support_v2` branch.

Would be great to get feedback.

Best regards,
       Oleh & Vitaly.

### CODE REVIEW

**Comment by cmhulbert** [comment]
Just wondering is this is still an active effort? I'd love to have Rust support for for Kaitai, as I'm trying to migrate some of my projects to Rust using existing kaitai structs

**Comment by Agile86** [comment]
Thanks for your interest. Yes, we are waiting for approval from maintainers.

**Comment by Agile86** [comment]
Updated PR with the latest changes from the master branch + small improvements.

File: shared/src/main/scala/io/kaitai/struct/languages/RustCompiler.scala [hunk 1]
```diff
@@ -1,612 +1,1391 @@
...
-    importList.add("std::default::Default")
-    importList.add("kaitai_struct::KaitaiStream")
-    importList.add("kaitai_struct::KaitaiStruct")
+    outHeader.puts("#![allow(unused_imports)]")
+    outHeader.puts("#![allow(non_snake_case)]")
+    outHeader.puts("#![allow(non_camel_case_types)]")
+    outHeader.puts("#![allow(irrefutable_let_patterns)]")
+    outHeader.puts("#![allow(unused_comparisons)]")
+    outHeader.puts("#![allow(arithmetic_overflow)]")
+    outHeader.puts("#![allow(overflowing_literals)]")
```
Comment (generalmimon) [inline shared/src/main/scala/io/kaitai/struct/languages/RustCompiler.scala:54]: These two suppressions hide legitimate errors, so I'll remove them.

**Review by generalmimon — COMMENTED** [review]
@Agile86 @revitalyr Thank you for your work, I'll now merge this.

I'm personally not very happy with the quality of the code (for example, too many mutable variables, too much string manipulation instead of modeling the necessary concepts in types in types and using [pattern matching](https://docs.scala-lang.org/tour/pattern-matching.html) like in the majority of KSC code), but it's better than what we have now.

Hopefully we can improve it over time. Going through the code now, it actually seems that it isn't bad overall, there are just a number of weird places that need rewriting. So when I have time in the future, I'll try to refactor these weird places one by one and then it might be actually good.

**Pull request merged**